kolibri/plugins/user_profile/assets/src/machines/changeFacilityMachine.js

```diff
@@ -71,238 +73,359 @@ const setMerging = assign({
+      SETCONTEXT: { actions: setInitialContext },
+      SELECTFACILITY: {
+        actions: assign({
+          targetFacility: (_, event) => {
+            return event.value;
+          },
+        }),
+        target: 'getFacilityUsernames',
       },
     },
-    checkUsernameExists: {
-      always: [
-        {
-          cond: context => !!context.accountExists,
-          target: 'usernameExists',
-          actions: setMerging,
-        },
-        {
-          target: 'confirmAccount',
-        },
-      ],
+  },
+  getFacilityUsernames: {
+    invoke: {
+      id: 'getUserNames',
+      src: (context, event) => connectToTargetKolibri(context, event),
+      onDone: {
+        target: 'selectFacility',
+        actions: checkExists,
+      },
+      onError: {
+        target: 'selectFacility',
+      },
     },
-    confirmAccount: {
-      meta: { route: 'CONFIRM_ACCOUNT', path: '/change_facility' },
-      on: {
-        NEW: {
-          cond: context => context.targetFacility && context.targetFacility.learner_can_sign_up,
-          target: 'createAccount',
-        },
-        CONTINUE: 'isAdmin',
-        BACK: 'changeFacility',
+  },
+  changeFacility: {
+    meta: { route: 'CONFIRM_CHANGE_FACILITY', path: '/change_facility' },
+    on: {
+      MERGE: {
+        target: 'mergeAccounts',
+        actions: [setMerging, send({ type: 'PUSH_HISTORY', value: 'changeFacility' })],
+      },
+      CONTINUE: {
+        target: 'checkUsernameExists',
+        actions: [send({ type: 'PUSH_HISTORY', value: 'changeFacility' })],
       },
     },
-    isAdmin: {
-      always: [
-        {
-          cond: context => context.role === 'superuser',
-          target: 'chooseAdmin',
-        },
-        {
-          target: 'checkIsMerging',
-        },
-      ],
+  },
+  checkUsernameExists: {
+    always: [
+      {
+        cond: context => !!context.accountExists,
+        target: 'usernameExists',
+      },
+      {
+        target: 'confirmAccountUsername',
+      },
+    ],
+  },
+  confirmAccountUsername: {
+    meta: { route: 'CONFIRM_ACCOUNT_USERNAME', path: '/change_facility' },
+    on: {
+      NEW: {
+        cond: context => context.targetFacility && context.targetFacility.learner_can_sign_up,
+        target: 'createAccount',
+        actions: [send({ type: 'PUSH_HISTORY', value: 'confirmAccountUsername' })],
+      },
+      CONTINUE: {
+        target: 'isAdmin',
+        actions: [send({ type: 'PUSH_HISTORY', value: 'confirmAccountUsername' })],
+      },
     },
-    chooseAdmin: {
-      meta: { route: 'CHOOSE_ADMIN', path: '/change_facility' },
-      on: {
-        CONTINUE: {
-          target: 'checkIsMerging',
-          cond: context => !!context.newSuperAdmin,
-        },
-        SELECTNEWSUPERADMIN: { actions: setNewSuperAdmin },
-        BACK: 'confirmAccount',
+  },
+  isAdmin: {
+    always: [
+      {
+        cond: context => context.role === 'superuser',
+        target: 'fetchSourceFacilityUsers',
```

Comment:
I see this is only used in the ChooseAdmin component. I've seen facilities with thousands of users. I don't feel comfortable having such list inside the machine context just to be used in one component. I think it'd be better if this is fetched and, later removed from the browser memory after the component is unmounted, only inside that component.

Reply:
Thank you, that's an important point.

I considered doing this inside the component, however, source facility users' data is not used only in the `chooseAdmin` state. Even before that, it's used in the machine in the `checkNeedsNewSuperAdmin` that decides whether the machine should go to the `chooseAdmin` or `checkIsMerging` state. In user experience, that translates to the decision whether the _Choose a new super admin_ page should be displayed or skipped, and for that, we need to have that data available before we mount the _Choose a new super admin_ page.

I agree that this should be resolved. I will think about it and let's chat together in a bit as we planned.

Reply:
With @jredrejo, we decided that for now, I will just make sure to clean that data. We also had some ideas on how to further optimize it, and because there are some product questions, I will raise this for discussion with the team if there are some more suggestions, it will most likely be implemented in the future as an improvement.

Regarding the data clean up, pushed a commit with that updated.